Src/WindowsServer/WindowsServer.Shared/Implementation/EnvironmentVariableMonitor.cs

```diff
@@ -0,0 +1,111 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    /// <summary>
+    /// Utility to monitor the value of environment variables which may change 
+    /// during the run of an application. Checks the environment variables 
+    /// intermittently.
+    /// </summary>
+    internal abstract class EnvironmentVariableMonitor : IDisposable
+    {
+        // Environment variables tracked by this monitor.
+        protected readonly ConcurrentDictionary<string, string> CheckedValues;
+
+        // enabled flag primarily used during dispose
+        protected volatile bool isEnabled = true;
+
+        // how often we allow the code to re-check the environment
+        protected TimeSpan checkInterval;
+
+        // timer object that will periodically update the environment variables
+        private readonly Timer environmentCheckTimer;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnvironmentVariableMonitor" /> class.
+        /// </summary>
+        public EnvironmentVariableMonitor(IEnumerable<string> envVars, TimeSpan checkInterval)
```

Comment:
Is there any advantage to pass the initial `envVars` to this constructor?
I mean, the monitor can build up its own list of monitored environment variables via the "GetOrAdd" within GetCurrentEnvironmentVariableValue.
So, why not simplify it and only monitor those variables that are actually queried via the public API.

Reply:
Oh, I guess the difference is in the "OnEnvironmentVariableUpdated" call-back. You can get a call-back even for variables you didn't actually call GetCurrentEnvironmentVariableValue on.

Reply:
Yes, I am optimizing for the Azure App Services scenario here.